src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -91,14 +95,16 @@ protected void updateStreamingIngestChannel(SnowflakeStreamingIngestChannel upda
   public void insertRecordToBuffer(SinkRecord record) {
     if (!hasChannelInitialized) {
       // This will only be called once at the beginning when an offset arrives for first time
-      // after connector starts/rebalance
-      init(record.kafkaOffset());
-      //            metricsJmxReporter.start();
+      // after connector starts
+      fetchLastCommittedOffsetToken();
+
       this.hasChannelInitialized = true;
     }
 
-    // ignore ingested files
-    if (record.kafkaOffset() > processedOffset.get()) {
+    // discard the record if the record offset is smaller or equal to server side offset, or if
+    // record is smaller than any other record offset we received before
+    if (record.kafkaOffset() > this.offsetPersistedInSnowflake.get()
+        && record.kafkaOffset() > processedOffset.get()) {
```

Comment:
Not sure if this is logged already but I think it would be good to capture what KC decides to use when it starts for a given topic partition in the event that we discover a bug in which we pick the wrong offset token.

Reply:
Yeah, good idea, let me log it in when hasChannelInitialized= false;